### DISCUSSION TITLE
How to make formula resolve git submodules?

### DISCUSSION THREAD

**Discussion by itsmattkc** [Homebrew/discussions#2100] [Writing Formulae/Casks] [answered]
I'm trying to write a formula that uses Git submodules to resolve dependencies. If I download a zip or tar.gz from GitHub, there's no `.git` folder so this submodule information is missing and CMake can't resolve them. If I set the `head` with a `tag`, then brew considers it a "head only"/unstable formula (despite the tag) and I get errors trying to make a tap out of it.

What are my options here? The only thing I can think of is to maintain/upload my own archive of the repo with the submodules already there, and that seems unnecessarily inconvenient.

**Comment by carlocab** [comment, marked as answer]
You can use a Git clone which automatically resolves submodules. You will need to specify both `tag` and `revision`. If the version information isn't recovered correctly from the tag you can also specify the `version` in the formula manually. For example:
```ruby
  url "https://github.com/grpc/grpc.git",
      tag:      "v1.39.1",
      revision: "2d6b8f61cfdd1c4d2d7c1aae65a4fbf00e3e0981"
  version "1391"
```
Note that Git clones are much slower and worse for caching than tarballs.

**Comment by nandahkrishna** [comment]
You can take a look at `arduino-cli` or other formulae which specify the stable URL with a `tag` and `revision`:
```ruby
url "https://github.com/arduino/arduino-cli.git",
    tag:      "0.18.3",
    revision: "d710b642ef7992a678053e9d68996c02f5863721"
```
This is less preferable and more expensive compared to using a tarball (which can be cached), but a fair few formulae do use this approach.

**Comment by itsmattkc** [comment]
Ah thanks, I didn't know the `url` could just be a Git repository.
